dali/test/dali_operator_test.h

```diff
@@ -23,30 +23,198 @@
+   *
+   * This is some sort or template method pattern.
+   */
+  using Verify = std::function<void(const std::vector<TensorListWrapper> & /* inputs */,
+                                    const std::vector<TensorListWrapper> & /* outputs */,
+                                    const Arguments &)>;
+
+
+  using VerifySingleIo = std::function<void(const TensorListWrapper & /* single input */,
+                                            const TensorListWrapper & /* single output */,
+                                            const Arguments &)>;
+
 
  protected:
-  void RunTest(const TensorListWrapper &input, const TensorListWrapper &output,
-               const Arguments &operator_arguments, const Verify &verify) {
+  /**
+   * Runs defined test.
+   *
+   * Call this method within GTest's testing macro (TEST_F, TEST_P, etc...).
+   * The method will set up a testing pipeline,
+   * push data through it and call output verification routines.
+   * Every RunTest call will set up its own pipeline.
+   *
+   * @tparam OutputBackend
+   * @param inputs all inputs to the pipeline
+   * @param outputs placeholder for outputs from pipeline
+   * @param operator_arguments arguments, with which the pipeline will be called
+   * @param verify function, that will be used for test verification
+   */
+  template<typename OutputBackend>
+  void
+  RunTest(const std::vector<TensorListWrapper> &inputs, std::vector<TensorListWrapper> &outputs,
+          const Arguments &operator_arguments, const Verify &verify) {
+    // TODO(mszolucha) implement
   }
 
 
-  void RunTest(const std::vector<TensorListWrapper> &inputs,
-               const std::vector<TensorListWrapper> &outputs,
-               const Arguments &operator_arguments, const std::vector<Verify> &verify) {
+  /**
+   * Convenient overload, for specific, single-input/single-output graph
+   */
+  template<typename OutputBackend>
+  void RunTest(const TensorListWrapper &input, TensorListWrapper &output,
+               const Arguments &operator_arguments, const VerifySingleIo &verify) {
+    std::string output_backend = detail::BackendStringName<OutputBackend>();
+    const auto batch_size = input.has_cpu() ? input.cpu().ntensor() : input.gpu().ntensor();
+    auto pipeline = CreatePipeline(batch_size, num_threads_);
+    if (input) {
+      AddInputToPipeline(*pipeline, input);
+    }
+    auto outputs = RunTestImpl<OutputBackend>(*pipeline, operator_arguments, input ? true : false,
+                                              input.backend(), output_backend);
+    verify(input, outputs[0], operator_arguments);
+    output = outputs[0];
   }
 
 
  private:
-  virtual GraphDescr GenerateOperatorsGraph() const noexcept = 0;
+  /**
+   * Template method. This function serves as a place of definition of
+   * operator graph, that will be tested within this fixture.
+   * @return tested operator graph
+   */
+  virtual GraphDescr GenerateOperatorGraph() const = 0;
 
 
   void SetUp() final {
```

Comment:
Why do you disallow SetUp() and TearDown() for child classes ?

Reply:
It can disturb proper testing. E.g. in case `DaliOperatorTest` would have some implementation of `SetUp` and `TearDown`, and subclass will override `SetUp`, `DaliOperatorTest::SetUp` will not be called. And since call is managed by GTest, I don't know if we can assume that calling
```
DerivedTest::SetUp() {
DaliOperatorTest::SetUp();
}
```
is safe